File: mobile/welcome-screen/ios-app/ShowUpWelcome/BasicsStep.swift
//  BasicsStep.swift
//  ShowUp · where in "The basics" the user is
//
//  The same shape as `FlowScreen.swift`, and for the same reasons: a typed position with no magic
//  values, an exhaustive switch at the host, and String raw values so `@SceneStorage` can persist it.
//
//  WHY THIS IS A SECOND ENUM AND NOT MORE CASES ON FlowScreen
//
//  FlowScreen is the app's outer position — sign-up, the tutorial cards, home. "The basics" is a
//  flow WITHIN one of those, with its own progress bar, its own back rules and its own persistence.
//  Folding four more cases into FlowScreen would put a tutorial card and a profile step in one list
//  whose order drives a screen transition, and the two do not share a transition.
//
//  Mirrors `profile/BasicsStep.kt`.

import Foundation

/// A step of "The basics".
enum BasicsStep: String, CaseIterable {
    case name
    /// The email address step.
    case email
    /// The code screen. **Not a fourth segment** — it holds at 2, see `progressSegment`.
    case emailVerify = "email_verify"
    case dob

    /// Which of the three segments is filled.
    ///
    /// `emailVerify` deliberately reports 2, the same as `email`: the user is on the email step
    /// until the code is confirmed, and a bar that advanced before that would claim progress they
    /// have not made. The flow README states it twice and the taxonomy's §2 registry agrees.
    var progressSegment: Int {
        switch self {
        case .name: return 1
        case .email, .emailVerify: return 2
        case .dob: return 3
        }
    }

    /// Whether the header shows a back chevron.
    ///
    /// False on `name` only. Profile creation is mandatory once entered — the user arrives from the
    /// tutorial and there is nothing behind step 1 — so the chevron is absent AND the swipe-back
    /// gesture is suppressed. A header with no chevron that the OS can still dismiss is worse than
    /// no rule at all.
    var hasBack: Bool { self != .name }

    /// `step_id` from the taxonomy's §2 vocabulary. The raw value IS the id.
    var stepId: String { rawValue }

    /// `step_index` for the funnel.
    ///
    /// Taken from the tickets, NOT the registry: `enums.json` §2 lists step_index as "—" for all
    /// four, so there is nothing to read. Recorded as a mismatch rather than silently invented.
    var stepIndex: Int { progressSegment }
}

/// Everything "The basics" collects, plus where the user got to.
///
/// One value so persistence is one write. The flow resumes at `step`, and resuming is silent: no
/// prompt, no toast, and no error state on arrival — which is why no error flag lives here. An
/// error belongs to a press, not to a saved position.
struct BasicsState: Equatable {
    var step: BasicsStep = .name
    var firstName: String = ""
    var email: String = ""
    var marketingConsent: Bool = false
}

File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/profile/BasicsStep.kt
/*
 * BasicsStep.kt
 * ShowUp · where in "The basics" the user is
 *
 * The same shape as `welcome/FlowScreen.kt`, and for the same reasons: a typed position with no
 * magic values, an exhaustive `when` at the host, and a Kotlin enum so `rememberSaveable` persists
 * it with nothing added.
 *
 * WHY THIS IS A SECOND ENUM AND NOT MORE CASES ON FlowScreen
 *
 * FlowScreen is the app's outer position -- sign-up, the tutorial cards, home. "The basics" is a
 * flow WITHIN one of those positions, with its own progress bar, its own back rules and its own
 * persistence. Folding four more cases into FlowScreen would put the tutorial's cards and a profile
 * step in one list whose ordinal drives a screen transition, and the two do not share a transition.
 *
 * That is also the answer to the navigation question this raises: nothing here needs a framework.
 * One linear flow, one enum, exhaustive. The trigger for revisiting that is written down in
 * docs/mobile-client-architecture-spike.md 1.5.1 and this is not it.
 */
package com.showup.profile

/** A step of "The basics". */
enum class BasicsStep {
    Name,
    Email,

    /**
     * The code screen. NOT a fourth segment.
     *
     * It carries [progressSegment] 2, the same as [Email]: the user is on the email step until the
     * code is confirmed, and the bar advancing before that would claim progress they have not made.
     * The flow README states it twice and the design's own §2 registry gives it `step_index` 2.
     */
    EmailVerify,

    Dob,
    ;

    /** Which of the three segments is filled. [EmailVerify] deliberately holds at 2. */
    val progressSegment: Int
        get() = when (this) {
            Name -> 1
            Email, EmailVerify -> 2
            Dob -> 3
        }

    /**
     * Whether the header shows a back chevron.
     *
     * False on [Name] only. Profile creation is mandatory once entered -- the user arrives from the
     * tutorial and there is nothing behind step 1 -- so the chevron is absent AND the platform back
     * is suppressed. A header with no chevron that the OS can still dismiss is worse than no rule.
     */
    val hasBack: Boolean get() = this != Name

    /** `step_id` from the taxonomy's §2 vocabulary. */
    val stepId: String
        get() = when (this) {
            Name -> "name"
            Email -> "email"
            EmailVerify -> "email_verify"
            Dob -> "dob"
        }

    /**
     * `step_index` for the funnel.
     *
     * Taken from the tickets, NOT from the registry: `enums.json` §2 lists step_index as "—" for
     * all four of these, so there is nothing to read. Recorded as a mismatch rather than silently
     * invented -- see the audit note in ProfileAnalytics.
     */
    val stepIndex: Int get() = progressSegment
}

/**
 * Everything "The basics" collects, plus where the user got to.
 *
 * Held as one value so persistence is one write. The flow resumes at [step], and resuming is
 * silent: no prompt, no toast, and no error state on arrival, which is why no error flag lives
 * here -- an error belongs to a press, not to the saved position.
 */
data class BasicsState(
    val step: BasicsStep = BasicsStep.Name,
    val firstName: String = "",
    val email: String = "",
    val marketingConsent: Boolean = false,
)
